### ISSUE TITLE
Extending schema.org 

### ISSUE THREAD

**Issue by marcolarosa** [describo/type-definitions#2]
## preamble

This repository has code to transform the schema.org jsonld (https://schema.org/version/latest/schemaorg-current-https.jsonld) into something Describo can consume. That is, it produces the internal profile that users get by default.

This ticket is about developing that concept by extending the schema.org definitions with useful ontologies/ descriptions.

## step 1 - extend schema.org with OLAC roles

As a first step, it would be good to get the roles defined in http://www.language-archives.org/REC/role.html into the definitions.
To do this, Describo itself can be used to create an RO-crate containing this content. By starting Describo (web or desktop) and then loading the profile `schema.org/vocabulary-creation-profile.json`, it's possible to define classes and properties that we can then join into the schema.org definition.

My feeling is that we should attach these properties to the `CreativeWork` class (https://schema.org/docs/full.html) which is a parent to both Dataset and MediaObject (which is renamed to File in ro-crate). That way, users will be able to use any of these roles when describing datasets and / or files or any other child of CreativeWork.

I've started this to test the idea and the ro-crate is @ https://github.com/describo/type-definitions/blob/master/schema.org-extensions/olac-roles.json. If you download this into a folder you can keep working on it. When all of the roles are described that file can be checked back in.

The key point to consider in this task but also any others we might undertake is making sensible decisions about what gets pulled in from other ontologies and where it gets plugged into. In this case, as these are contributor role definitions, then it makes sense (I think!) to attach them at the level of CreativeWork.

> You might want to restructure that extensions folder so it contains folders (olac-roles, ro-crate-extensions) that each contain an ro-crate-metadata file to make updating the crates easier. That is, the folder can just be loaded in and edited without having to create temp folders and copying files around.

## step 2 - revise the type definition creation

Once this is done we can then update the generation of the type definitions. We want this script to:
* load all of the schema.org classes - probably into a map keyed on the class name (e.g. CreativeWork)
* join all of the relevant properties into their respective classes
* iterate over the extensions in the `schema.org-extensions` folder and add any new classes
  * > New classes should be joined in where they make sense. That is, if we were adding a class (for example) Truck an appropriate parent would be https://schema.org/Vehicle rather than https://schema.org/Thing. Not relevant for this example but something to keep in mind when creating the extensions crate

  * then join any properties that have been defined
* finally write out a type definitions file containing an object of entity definitions keyed on the class name - see snippet following:

```JSON
{  
    ...,
    "Restaurant": {
        "id": "http://schema.org/Restaurant",
        "name": "Restaurant",
        "help": "A restaurant.",
        "subClassOf": ["FoodEstablishment"],
        "allowAdditionalProperties": false,
        "inputs": [],
        "linksTo": [],
        "hierarchy": [
            "Restaurant",
            "FoodEstablishment",
            "LocalBusiness",
            "Organization",
            "Place",
            "Thing"
        ]
    },
    "LocalBusiness": {
        "id": "http://schema.org/LocalBusiness",
        "name": "LocalBusiness",
        "help": "A particular physical business or branch of an organization. Examples of LocalBusiness include a restaurant, a particular branch of a restaurant chain, a branch of a bank, a medical practice, a club, a bowling alley, etc.",
        "subClassOf": ["Organization", "Place"],
        "allowAdditionalProperties": false,
        "inputs": [
            {
                "id": "http://schema.org/branchOf",
                "name": "branchOf",
                "help": "The larger organization that this local business is a branch of, if any. Not to be confused with (anatomical)[[branch]].",
                "multiple": true,
                "type": ["Organization"]
            },
            {
                "id": "http://schema.org/currenciesAccepted",
                "name": "currenciesAccepted",
                "help": "The currency accepted.\\n\\nUse standard formats: [ISO 4217 currency format](http://en.wikipedia.org/wiki/ISO_4217) e.g. \"USD\"; [Ticker symbol](https://en.wikipedia.org/wiki/List_of_cryptocurrencies) for cryptocurrencies e.g. \"BTC\"; well known names for [Local Exchange Tradings Systems](https://en.wikipedia.org/wiki/Local_exchange_trading_system) (LETS) and other currency types e.g. \"Ithaca HOUR\".",
                "multiple": true,
                "type": ["Text"]
            },
            {
                "id": "http://schema.org/openingHours",
                "name": "openingHours",
                "help": "The general opening hours for a business. Opening hours can be specified as a weekly time range, starting with days, then times per day. Multiple days can be listed with commas ',' separating each day. Day or time ranges are specified using a hyphen '-'.\\n\\n* Days are specified using the following two-letter combinations: ```Mo```, ```Tu```, ```We```, ```Th```, ```Fr```, ```Sa```, ```Su```.\\n* Times are specified using 24:00 format. For example, 3pm is specified as ```15:00```, 10am as ```10:00```. \\n* Here is an example: <code>&lt;time itemprop=\"openingHours\" datetime=&quot;Tu,Th 16:00-20:00&quot;&gt;Tuesdays and Thursdays 4-8pm&lt;/time&gt;</code>.\\n* If a business is open 7 days a week, then it can be specified as <code>&lt;time itemprop=&quot;openingHours&quot; datetime=&quot;Mo-Su&quot;&gt;Monday through Sunday, all day&lt;/time&gt;</code>.",
                "multiple": true,
                "type": ["Text"]
            },
            {
                "id": "http://schema.org/paymentAccepted",
                "name": "paymentAccepted",
                "help": "Cash, Credit Card, Cryptocurrency, Local Exchange Tradings System, etc.",
                "multiple": true,
                "type": ["Text"]
            },
            {
                "id": "http://schema.org/priceRange",
                "name": "priceRange",
                "help": "The price range of the business, for example ```$$$```.",
                "multiple": true,
                "type": ["Text"]
            }
        ],
        "linksTo": ["SpecialAnnouncement"],
        "hierarchy": ["LocalBusiness", "Organization", "Place", "Thing"]
    },
    ...,
```

To explain this structure:

```JSON
 "Restaurant": {
        "id": "http://schema.org/Restaurant",
        "name": "Restaurant",
        "help": "A restaurant.",
        "subClassOf": ["FoodEstablishment"],
        "allowAdditionalProperties": false,
        "inputs": [],
        "linksTo": [],
        "hierarchy": [
            "Restaurant",
            "FoodEstablishment",
            "LocalBusiness",
            "Organization",
            "Place",
            "Thing"
        ]
    },
```

Consider the definition @ https://schema.org/Restaurant and note that it's a child in two different hierarchies

* The key is the class name
* The object definition is:
  * id: the URI to the definition of the thing,
  * name: the short name,
  * help: the description,
  * subClassOf: a link to its parent class,
  * showAdditionalProperties: unused,
  * inputs: an array of inputs defined on this class specifically - look at the localBusiness definition for an example,
  * linksTo: unused (I think... it's been a while),
  * hierarchy: the class hierarchy - perhaps the most important property as it tells describo how to find the properties that can be defined for this entity. Note that it captures both of the hierarchies.

## step 3 - possible future work

This work basically paves the way for extending schema.org with any number of existing ontologies and definitions. RO-Crate already references a number of ontologies in a general sense but the specific properties and classes are not known to describo. Depending on how we go with this first task, we might find an easy, programmatic way to get all of these ontologies and others in for use.

![Screenshot 2024-01-12 at 1 28 41 pm](https://github.com/describo/type-definitions/assets/2639995/bec486ab-18fa-4308-9e47-b5b5434ffc85)
